Code/Framework/AzToolsFramework/AzToolsFramework/Entity/PrefabEditorEntityOwnershipService.cpp

```diff
@@ -104,39 +105,62 @@ namespace AzToolsFramework
+        if (currentParentId.IsValid())
+        {
+            auto* instanceEntityMapperInterface = AZ::Interface<Prefab::InstanceEntityMapperInterface>::Get();
+            if (instanceEntityMapperInterface)
+            {
+                newOwningInstance = instanceEntityMapperInterface->FindOwningInstance(currentParentId);
+            }
+        }
+
+        // If no parent was set, the mapper can't be found, or the parent has no owning instance, add to the current focus.
+        if (!newOwningInstance.has_value())
+        {
+            auto* prefabFocusInterface = AZ::Interface<Prefab::PrefabFocusInterface>::Get();
+            if (prefabFocusInterface)
+            {
+                newOwningInstance = prefabFocusInterface->GetFocusedPrefabInstance(m_entityContextId);
+            }
+        }
+
+        // If the focus interface isn't handled, fall back to the root instance
+        if (!newOwningInstance.has_value())
+        {
+            newOwningInstance = *m_rootInstance;
+        }
+
+        // Setup undo node
+        ScopedUndoBatch undoBatch("Add entity");
         Prefab::PrefabDom instanceDomBeforeUpdate;
-        Prefab::PrefabDomUtils::StoreInstanceInPrefabDom(*m_rootInstance, instanceDomBeforeUpdate);
+        Prefab::PrefabDomUtils::StoreInstanceInPrefabDom(newOwningInstance->get(), instanceDomBeforeUpdate);
 
-        m_rootInstance->AddEntity(*entity);
+        newOwningInstance->get().AddEntity(*entity);
         HandleEntitiesAdded({ entity });
-        AZ::TransformBus::Event(entity->GetId(), &AZ::TransformInterface::SetParent, m_rootInstance->m_containerEntity->GetId());
+        AZ::TransformBus::Event(entity->GetId(), &AZ::TransformInterface::SetParent, newOwningInstance->get().GetContainerEntityId());
 
         Prefab::PrefabUndoHelpers::UpdatePrefabInstance(
-            *m_rootInstance, "Undo adding entity", instanceDomBeforeUpdate, undoBatch.GetUndoBatch());
+            newOwningInstance->get(), "Add entity", instanceDomBeforeUpdate, undoBatch.GetUndoBatch());
     }
 
     void PrefabEditorEntityOwnershipService::AddEntities(const EntityList& entities)
     {
         AZ_Assert(IsInitialized(), "Tried to add entities without initializing the Entity Ownership Service");
-        ScopedUndoBatch undoBatch("Undo adding entities");
-        Prefab::PrefabDom instanceDomBeforeUpdate;
-        Prefab::PrefabDomUtils::StoreInstanceInPrefabDom(*m_rootInstance, instanceDomBeforeUpdate);
 
-        for (AZ::Entity* entity : entities)
-        {
-            m_rootInstance->AddEntity(*entity);
-        }
-
-        HandleEntitiesAdded(entities);
+        // Group the undo nodes under a single batch.
+        ScopedUndoBatch undoBatch("Add entities");
 
         for (AZ::Entity* entity : entities)
         {
-            AZ::TransformBus::Event(entity->GetId(), &AZ::TransformInterface::SetParent, m_rootInstance->m_containerEntity->GetId());
+            AddEntity(entity);
```

Comment:
The scoped undo batch node that is created at the beginning of the function should actually wrap all the created nodes into one. I don't think I can easily reduce the amount of nodes created (each entity could belong to a different owning instance).

Reply:
But why would someone add an entity to a prefab instance and then call PrefabEOS::AddEntity(). That doesn't sound like a valid workflow we should be supporting. General order of flow should be 'PrefabPublicHandler/OtherSystems' -> PrefabEOS->PrefabSystem

So, I think you can remove the part about 'FindOwningInstance' altogether I think. We can keep the flow simple like 'If you add an entity to the EOS, it'll be added to the current focused prefab or the root prefab.'

Reply:
The API uses raw pointers to entities because that's what the entity context supported too(EditorEntityContext::AddEntity() also eventually comes to PrefabEditorEOS). Changing it to create an entity instead may not be what different systems want. 

But for systems that call the entitycontext or eos directly, it's rare that it would be belonging to a prior prefab instance. Those usecases would look like "I created this entity with the required components I want. Now add this to the entity context". So, rather than covering the possibility of entities belonging to prior instance, it should be treated as an outlier. If an entity belongs to a prior instance, throwing an error and returning would be better because we don't want to encourage that behavior.

Reply:
Makes sense. I can handle that behavior and flip the logic so that all entities are added to the same instance then.

Reply:
Ok, I'm actually revisiting this, and there's a small wrinkle we missed. The check I added here verifies the owning instance of the parent, not of the entity itself.
```
        Prefab::InstanceOptionalReference newOwningInstance;
        // If the entity has a parent set, try to find that entity's owning instance and use that.
        if (currentParentId.IsValid())
        {
            auto* instanceEntityMapperInterface = AZ::Interface<Prefab::InstanceEntityMapperInterface>::Get();
            if (instanceEntityMapperInterface)
            {
                newOwningInstance = instanceEntityMapperInterface->FindOwningInstance(currentParentId);
            }
        }
```
I think this is still valid honestly? If an entity is created as a child of an existing registered entity, and then it's added to the EOS, shouldn't it go under the parent's instance? Else we could get in the case where an entity is associated to a nested prefab instance while its child is actually part of the root, which is an invalid case.

Should we maybe enforce that the entity should not have a parent when it's added to the EOS? But then if the parent is changed programmatically it would not trigger reparenting (since that's on the editor side) so we're still potentially in a broken state.

Reply:
Previously, the entity was always being parented to the root container entity
```
AZ::TransformBus::Event(entity->GetId(), &AZ::TransformInterface::SetParent, m_rootInstance->m_containerEntity->GetId())
```
It means that even if they set the parent themselves, it was always be parented to the root container entity. When the EOS gets the entity, it's not even guaranteed that it will have a tranform component. If the users want to retain the parent, they can call the PrefabPublicHandler to create an entity with a parent instead. Hopefully in the future, we can provide a unified way of adding entities. But the EOS was intended to be a lightweight system to register entities with the underlying system(slices/prefabs). Rather than try to prepare the api for different possible scenarios, I suggest keeping it as a lightweight api that does the required.

Besides,

> If an entity is created as a child of an existing registered entity, and then it's added to the EOS, shouldn't it go under the parent's instance?

this will be circumventing the focus mode which makes it even confusing to the caller of the eos api to know how to prepare the entities, which will retain parents, which will be in focus, etc. So, always reparenting to the root/focused prefab sounds like an iterative step we can take for now. If a use case comes when someone would like to call EOS and retain the parent info too, we can revisit this then.

Reply:
Ok, so we always override the parent to the container of the focused prefab. I think that solves all my issues actually :D

Thanks for clarifying, I'll make the changes and then we can re-verify if it makes sense :)